osquery/tables/system/windows/certificates.cpp

```diff
@@ -307,7 +337,236 @@ void parseSystemStoreString(LPCWSTR sysStoreW,
   }
 }
 
-/// Enumerate and process a certificate store
+#pragma pack(push, 1)
+struct Header {
+  unsigned long propid;
+  unsigned long unknown;
+  unsigned long size;
+};
+#pragma pack(pop)
+
+/**
+ * This function extracts an encoded certificate from a proprietary Windows
+ * file format, which is described in the links below. Briefly, the file is
+ * an array of elements where each element contains a header followed by a
+ * variable length data buffer. The encoded certificate is stored in the data
+ * buffer of one of these elements whose header has a specific `propid` field.
+ *
+ * Links:
+ * https://itsme.home.xs4all.nl/projects/xda/smartphone-certificates.html
+ * https://github.com/wine-mirror/wine/blob/f9301c2b66450a1cdd986e9052fcaa76535ba8b7/dlls/crypt32/crypt32_private.h#L146
+ */
+Status getEncodedCert(std::basic_istream<BYTE>& blob,
+                      std::vector<BYTE>& encodedCert) {
+  static const unsigned long CERT_CERT_PROP_ID = 0x20;
+
+  Header hdr;
+
+  while (true) {
+    blob.read(reinterpret_cast<BYTE*>(&hdr), sizeof(hdr));
+    if (!blob.good()) {
+      return Status::failure("Malformed certificate blob");
+    }
+
+    if (hdr.propid != CERT_CERT_PROP_ID) {
+      blob.ignore(hdr.size);
+      if (!blob.good()) {
+        return Status::failure("Malformed certificate blob");
+      }
+      continue;
+    }
+
+    encodedCert.resize(hdr.size);
+    blob.read(encodedCert.data(), hdr.size);
+    if (!blob.good()) {
+      return Status::failure("EOF in certificate blob when reading data");
+    }
+    break;
+  }
+
+  return Status::success();
+}
+
+void addCertRow(PCCERT_CONTEXT certContext,
+                const std::string& storeId,
+                const std::string& sid,
+                const std::string& storeName,
+                const std::string& username,
+                const std::string& storeLocation,
+                QueryData& results) {
+  std::vector<char> certBuff;
+  getCertCtxProp(certContext, CERT_HASH_PROP_ID, certBuff);
+  std::string fingerprint;
+  toHexStr(certBuff.begin(), certBuff.end(), fingerprint);
+
+  Row r;
+  r["sid"] = sid;
+  r["username"] = username;
+  r["store_id"] = storeId;
+  r["sha1"] = fingerprint;
+  certBuff.resize(256, 0);
+  std::fill(certBuff.begin(), certBuff.end(), 0);
+  CertGetNameString(certContext,
+                    CERT_NAME_SIMPLE_DISPLAY_TYPE,
+                    0,
+                    nullptr,
+                    certBuff.data(),
+                    static_cast<unsigned long>(certBuff.size()));
+  r["common_name"] = certBuff.data();
+  VLOG(1) << "cert name: " << certBuff.data();
```

Comment:
I don't like this `VLOG`. If you are adding these for your own purposes of developing the table then we should remove them before landing. I don't see the value-add for a user running the table and attempting to debug.

Reply:
done 👍